### PR TITLE
[Improve][Transform] Sql transform support inner strucy query

### LINKED ISSUE #6468: [Improve][SQL-Transform] improve sql transform to support struct query like row_a.inner_row_b.column_c
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

Now the sql transform doesn't support struct query. when i run `row_a.inner_row_b.column_c`, it will be recognized as table `row_a.inner_row_b`, colum `column_c`.  
so it will got `column_c` not exist error.

I will improve the current SQL transform logical to support the struct query.

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by liunaijie**
### Purpose of this pull request
close #6468 

Now the struct query will be recognized as `table_name.column_name`, I am not change this because i think there are someone still use this query pattern. just add a backfill logical to handle it.
For map stuct, it should be the latest struct.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

add e2e test

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Review by hailin0 — COMMENTED** [review]
Thanks for your contribution, need to update the documentation?

**Comment by hailin0** [comment]
LGTM

waiting for ci passed

**Comment by liunaijie** [comment]
> Thanks for your contribution, need to update the documentation?

Updated

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/java/org/apache/seatunnel/e2e/transform/TestSQLIT.java
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/java/org/apache/seatunnel/e2e/transform/TestSQLIT.java]: Hi @liunaijie . Why disabled on spark?

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java
Comment (Hisoka-X) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java]: Can you add a new `indexOf` method in `SeaTunnelRowType` which no IllegalArgumentException exception is thrown. Then use the result of the index to determine whether there is a corresponding fieldname. Because data processing is performed through try catch at now, the performance is particularly bad.

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLType.java
Comment (Hisoka-X) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLType.java]: ditto

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/java/org/apache/seatunnel/e2e/transform/TestSQLIT.java
Comment (liunaijie) [reply seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/java/org/apache/seatunnel/e2e/transform/TestSQLIT.java]: image
https://[productionresultssa14.blob.core.windows.net/actions-results/f4b0ae05-af44-4261-9d26-5c580da07424/workflow-job-run-8ea7adc3-2928-51c7-be22-ae070f910ac1/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-14T04%3A08%3A17Z&sig=l03NcLlmGAPq0MGARzURtW7BP2thnPwOjWZ543qA%2FNE%3D&sp=r&spr=https&sr=b&st=2024-03-14T03%3A58%3A12Z&sv=2021-12-02](https://productionresultssa14.blob.core.windows.net/actions-results/f4b0ae05-af44-4261-9d26-5c580da07424/workflow-job-run-8ea7adc3-2928-51c7-be22-ae070f910ac1/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-14T04%3A08%3A17Z&sig=l03NcLlmGAPq0MGARzURtW7BP2thnPwOjWZ543qA%2FNE%3D&sp=r&spr=https&sr=b&st=2024-03-14T03%3A58%3A12Z&sv=2021-12-02)

i meet an exception on spark engine, so i disable it. will check the failed reason

Comment (Hisoka-X) [reply seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/java/org/apache/seatunnel/e2e/transform/TestSQLIT.java]: Seem like the spark translation have some problem when use map type. You can create an issue and add it into your disabledReason.

**Review by Hisoka-X — APPROVED** [review]
LGTM if ci passed

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java
Comment (zhilinli123) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:228]:
```
  public int indexOf(String fieldName, boolean throwExceptionWhenNotFound) {
        for (int i = 0; i < fieldNames.length; i++) {
            if (fieldNames[i].equals(fieldName)) {
                return i;
            }
        }
        if (throwExceptionWhenNotFound) {
            throw new IllegalArgumentException(String.format("can't find field [%s]", fieldName));
        } else {
            return -1;
        }
    }
``` Logical conflict？
cc  @hailin0 @Hisoka-X

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java
Comment (zhilinli123) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:226]:
```
  public int indexOf(String fieldName, boolean throwExceptionWhenNotFound) {
        for (int i = 0; i < fieldNames.length; i++) {
            if (fieldNames[i].equals(fieldName)) {
                return i;
            }
        }
        if (throwExceptionWhenNotFound) {
            throw new IllegalArgumentException(String.format("can't find field [%s]", fieldName));
        } else {
            return -1;
        }
    }
``` Logical conflict？
cc  @hailin0 @Hisoka-X

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java
Comment (zhilinli123) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:228]: cc @Hisoka-X @hailin0

Comment (liunaijie) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:228]: check the field name is in this struct. if not exist, choose throw exception or return -1 index let's the caller handle.

Comment (zhilinli123) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:228]: image
Repetitive code logic

Comment (Hisoka-X) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:228]: Thanks @zhilinli123 for ping me. I think it's different. `indexOf` only throw sub fieldname. The `fullyQualifiedName` and `columnNames[i]` not equals. Like field `user.age`, we should throw `user.age` not `age`.

Comment (hailin0) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLFunction.java:228]: i agree with @Hisoka-X

**Review by hailin0 — APPROVED** [review]

**Review by zhilinli123 — APPROVED** [review]

**Pull request merged**
